request_test.go

```diff
@@ -247,3 +249,108 @@ func TestOpendirHandleReuse(t *testing.T) {
 	rpkt = request.call(handlers, pkt, nil, 0)
 	assert.IsType(t, &sshFxpNamePacket{}, rpkt)
 }
+
+func Test_toLocalPath(t *testing.T) {
+	type args struct {
+		workDir string
+		p       string
+	}
+	tests := []struct {
+		name string
+		goos string
+		args args
+		want string
+	}{
+		{
+			name: "empty path",
+			goos: "linux",
+			args: args{p: ""},
+			want: "",
+		},
+		{
+			name: "relative path",
+			goos: "linux",
+			args: args{p: "file"},
+			want: "file",
+		},
+		{
+			name: "absolute path",
+			goos: "linux",
+			args: args{p: "/file"},
+			want: "/file",
+		},
+		{
+			name: "empty path",
+			goos: "linux",
+			args: args{workDir: "/home/user", p: ""},
+			want: "/home/user",
+		},
+		{
+			name: "relative path",
+			goos: "linux",
+			args: args{workDir: "/home/user", p: "file"},
+			want: "/home/user/file",
+		},
+		{
+			name: "relative path with .",
+			goos: "linux",
+			args: args{workDir: "/home/user", p: "."},
+			want: "/home/user",
+		},
+		{
+			name: "absolute path",
+			goos: "linux",
+			args: args{workDir: "/home/user", p: "/file"},
+			want: "/file",
+		},
+		{
+			name: "empty path",
+			goos: "windows",
+			args: args{workDir: "C:\\Users\\User", p: ""},
+			want: "C:\\Users\\User",
+		},
+		{
+			name: "relative path",
+			goos: "windows",
+			args: args{workDir: "C:\\Users\\User", p: "file"},
+			want: "C:\\Users\\User\\file",
+		},
+		{
+			name: "relative path with .",
+			goos: "windows",
+			args: args{workDir: "C:\\Users\\User", p: "."},
+			want: "C:\\Users\\User",
+		},
+		{
+			name: "absolute path",
+			goos: "windows",
+			args: args{workDir: "C:\\Users\\User", p: "C:\\file"},
+			want: "C:\\file",
+		},
+		{
+			name: "relative unix-like path",
+			goos: "windows",
+			args: args{workDir: "C:\\Users\\User", p: "./file"},
+			want: "C:\\Users\\User\\file",
+		},
+		{
+			name: "absolute unix-like path",
+			goos: "windows",
+			args: args{workDir: "C:\\Users\\User", p: "/C:/file"},
+			want: "C:\\file",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(fmt.Sprintf("%s %s %s", tt.goos, tt.args.workDir, tt.name), func(t *testing.T) {
+			if runtime.GOOS != tt.goos {
+				t.Skipf("Skipping test for %s on %s", tt.goos, runtime.GOOS)
+			}
+
+			var s Server
+			if tt.args.workDir != "" {
+				_ = WithServerWorkingDirectory(tt.args.workDir)(&s)
+			}
+			assert.Equal(t, tt.want, toLocalPath(s.workDir, tt.args.p), "wrong local path")
```

Comment:
I’m not sure what utility there is in instantiating a whole `Server` when we could just pass `tt.args.workDir` directly into `toLocalPath`.

Reply:
The idea was to test the newly introduced API, if the implementation of `WithServerWorkingDirectory` ever changes, the test will catch the change in behavior. I can simplify.

Reply:
If you want to test the newly introduced API, then you need to actually test to make sure that a request made to that server is interpreted into the correct path.

All you’re testing here is that `s.workDir` is set by `WithServerWorkingDirectory`.

Reply:
I think the nuance of what I was going for is different, I wasn't so much testing that `server.workDir` gets set correctly, I was testing that the input a user would use for `WithServerWorkingDirectory` produces the right output from `toLocalPath`. So in that sense it tested exactly what I wanted, but I understand you feel it's not the best approach. Also, my naming was lazy, I shouldn't have put `workDir` under `tt.args` to better reflect the intent.

I'll fix up my tests according to your input :+1:.

Reply:
I understood what you were trying for, but there was no guarantee that `toLocalPath` is getting called with `server.workDir`. That part of the loop has to be closed manually in the function calls.

Actually… thinking about it, if we made `toLocalPath` a receiver method off of `Server` then this test would be the correct way to test it. It also closes the loop mentioned above by making the `WithServerWorkingDirectory` argument inherent to `toLocalPath` rather than explicit.